Migrate AWS and EKS providers to AWS SDK v2 for Go

embik opened the pull request:
**What this PR does / why we need it**:

This PR migrates both the AWS and EKS code to the v2 of AWS SDK for Go. Some light API changes that seemed to rely on the internal details of the AWS SDK are adjusted. There's not really a reason to have a string pointer when you can have just a string.

The `Verifier` part of the `aws-iam-authenticator` code has been removed because it relied on partition information no longer exposed by the SDK. In addition, this code was unused in KKP, so removing it seemed the cleanest option.

**Which issue(s) this PR fixes**:
<!--optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged-->
Fixes #10841

**What type of PR is this?**
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:
<!--
Write your release note. Release notes are being used to generate the changelog:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
The AWS and EKS cloud provider implementations now use the AWS Go SDK v2 to communicate with AWS
```

**Documentation**:
<!--
Please do one of the following options:
- Add a link to the existing documentation
- Add a link to the kubermatic/docs pull request
- If no documentation change is applicable then add:
  - TBD (documentation will be added later)
  - NONE (no documentation needed for this PR)
-->
```documentation
NONE
```

embik committed:
Initial migration to AWS SDK v2

Signed-off-by: Marvin Beckers <marvin@kubermatic.com>

embik committed:
Port authenticator code to aws-sdk-go-v2

Signed-off-by: Marvin Beckers <marvin@kubermatic.com>

embik committed:
Fix last compile issues

Signed-off-by: Marvin Beckers <marvin@kubermatic.com>

embik committed:
Update codegen

Signed-off-by: Marvin Beckers <marvin@kubermatic.com>

embik committed:
Fix import statement ordering

Signed-off-by: Marvin Beckers <marvin@kubermatic.com>

embik committed:
Fix linting issues

Signed-off-by: Marvin Beckers <marvin@kubermatic.com>

embik committed:
Prevent nil pointer panic

Signed-off-by: Marvin Beckers <marvin@kubermatic.com>

embik committed:
Fix missed lint issue

Signed-off-by: Marvin Beckers <marvin@kubermatic.com>

embik committed:
Make permission formatting npe-proof

Signed-off-by: Marvin Beckers <marvin@kubermatic.com>

embik committed:
Next attempt at fixing integration tests

Signed-off-by: Marvin Beckers <marvin@kubermatic.com>

embik committed:
Write a custom compare function for ip permissions

Signed-off-by: Marvin Beckers <marvin@kubermatic.com>

embik committed:
Fix import statement

Signed-off-by: Marvin Beckers <marvin@kubermatic.com>

xrstf committed:
post-rebase blues

ahmedwaleedmalik reviewed (approved):
Great job, especially with the context propagation.

/approve
/hold if @harshita-sharma012 wants to take a look as well

imharshita commented on pkg/provider/cloud/eks/provider.go:
```diff
@@ -432,66 +417,61 @@ func ListUpgrades(ctx context.Context,
-			return utilerrors.New(reqErr.StatusCode(), reqErr.Message())
-		}
-		return errors.New(aerr.Message())
+		return utilerrors.New(500, fmt.Sprintf("%s: %s", aerr.ErrorCode(), aerr.ErrorMessage()))
```
why hardcoded status code to 500?

xrstf replied:
Good catch! I think we should get rid of this error handling code alltogether, to quote my commit:

> smithy.APIError implementations all seem to be doing the same, if not better string marshalling that we would do.
>
> Also using HTTP 500 (internal server error) for every possible fault is not ideal (ValidateCredentials() should not error out with an internal server error on bad credentials), and in general the code in the provider package should not return HTTP-related things, as controllers also use it.
>
> I removed the DecodeError() entirely, as it seems to serve no purpose anymore with SDK v2. :)

imharshita replied:
but, this part of code from external cluster controller expects error to be an `utilerrors.HTTPError` :
https://github.com/kubermatic/kubermatic/blob/master/pkg/controller/master-controller-manager/external-cluster/cluster_controller.go#L106

either we can modify controller here, or I can create a separate pr as this will not immediately affect the controller unless `notFound` etc. clusters

embik replied:
@imharshita can you please take another look, Christoph fixed this in https://github.com/kubermatic/kubermatic/pull/10922/commits/628c8727b937b7eb338b1eb009283f3ca61d56da I think.

embik replied:
@imharshita I reworked the DecodeError function, it should now return what is expected by the controller code, if that information was included in the returned error. Please take another look.

imharshita replied:
@embik  regarding hardcoding 500 code in DecodeError , I will fix that in a separate pr.

embik replied:
the hardcoded 500 error is already resolved. We only fall back to 500 if no HTTP information was included in the error.

xrstf committed:
remove error decoding

smithy.APIError implementations all seem to be doing the same, if not
better string marshalling that we would do.

Also using HTTP 500 (internal server error) for every possible fault
is not ideal (ValidateCredentials() should not error out with an internal
server error on bad credentials), and in general the code in the
provider package should not return HTTP-related things, as controllers
also use it.

I removed the DecodeError() entirely, as it seems to serve no purpose
anymore with SDK v2. :)

embik committed:
Only read from array if length is 2 and set the correct option

Signed-off-by: Marvin Beckers <marvin@kubermatic.com>

embik committed:
Revert "remove error decoding"

This reverts commit 628c8727b937b7eb338b1eb009283f3ca61d56da.

embik committed:
Rework DecodeError to include HTTP status codes

Signed-off-by: Marvin Beckers <marvin@kubermatic.com>

embik committed:
Make linter happy

Signed-off-by: Marvin Beckers <marvin@kubermatic.com>
